plan/join_reorder.go

```diff
@@ -0,0 +1,205 @@
+func (e *joinReOrderSolver) reorderJoin(group []LogicalPlan, conds []expression.Expression) {
+	e.graph = make([]edgeList, len(group))
+	e.group = group
+	e.visited = make([]bool, len(group))
+	e.resultJoin = nil
+	e.groupRank = make([]*rankInfo, len(group))
+	for i := 0; i < len(e.groupRank); i++ {
+		e.groupRank[i] = &rankInfo{
+			nodeID: i,
+			rate:   1.0,
+		}
+	}
+	for _, cond := range conds {
+		if f, ok := cond.(*expression.ScalarFunction); ok {
+			if f.FuncName.L == ast.EQ {
+				lCol, lok := f.Args[0].(*expression.Column)
+				rCol, rok := f.Args[1].(*expression.Column)
+				if lok && rok {
+					lID := findColumnIndexByGroup(group, lCol)
+					rID := findColumnIndexByGroup(group, rCol)
+					if lID != rID {
+						e.graph[lID] = append(e.graph[lID], &rankInfo{nodeID: rID})
+						e.graph[rID] = append(e.graph[rID], &rankInfo{nodeID: lID})
+						continue
+					}
+				}
+			}
+			id := -1
+			rate := 1.0
+			cols, _ := extractColumn(f, nil, nil)
+			for _, col := range cols {
+				idx := findColumnIndexByGroup(group, col)
+				if id == -1 {
+					switch f.FuncName.L {
+					case ast.EQ:
+						rate *= 0.1
+					case ast.LT, ast.LE, ast.GE, ast.GT:
+						rate *= 0.3
+					default:
+						rate *= 0.9
+					}
+					id = idx
+				} else {
+					id = -1
+					break
+				}
+			}
+			if id != -1 {
+				e.groupRank[id].rate *= rate
+			}
+		}
+	}
+	for _, edge := range e.graph {
+		for _, node := range edge {
+			node.rate = e.groupRank[node.nodeID].rate
+		}
+	}
+	sort.Sort(e)
+	for _, edge := range e.graph {
+		sort.Sort(edge)
+	}
+	var cartesianJoinGroup []LogicalPlan
+	for j := 0; j < len(e.groupRank); j++ {
+		i := e.groupRank[j].nodeID
+		if !e.visited[i] {
+			e.resultJoin = e.group[i]
+			e.walkGraphAndComposeJoin(i)
+			cartesianJoinGroup = append(cartesianJoinGroup, e.resultJoin)
+		}
+	}
+	e.makeBushyJoin(cartesianJoinGroup)
+}
+
+// Make cartesian join as bushy tree.
+func (e *joinReOrderSolver) makeBushyJoin(cartesianJoinGroup []LogicalPlan) {
+	for len(cartesianJoinGroup) > 1 {
+		resultJoinGroup := make([]LogicalPlan, 0, len(cartesianJoinGroup))
+		for i := 0; i < len(cartesianJoinGroup); i += 2 {
+			if i+1 == len(cartesianJoinGroup) {
+				resultJoinGroup = append(resultJoinGroup, cartesianJoinGroup[i])
+				break
+			}
+			resultJoinGroup = append(resultJoinGroup, e.newJoin(cartesianJoinGroup[i], cartesianJoinGroup[i+1]))
+		}
+		cartesianJoinGroup = resultJoinGroup
+	}
+	e.resultJoin = cartesianJoinGroup[0]
+}
+
+func (e *joinReOrderSolver) newJoin(lChild, rChild LogicalPlan) *Join {
+	join := &Join{
+		JoinType:        InnerJoin,
+		reordered:       true,
+		baseLogicalPlan: newBaseLogicalPlan(Jn, e.allocator),
+	}
+	join.initID()
+	join.SetChildren(lChild, rChild)
+	join.SetSchema(append(lChild.GetSchema().DeepCopy(), rChild.GetSchema().DeepCopy()...))
+	lChild.SetParents(join)
+	rChild.SetParents(join)
+	return join
+}
+
+// walkGraph implement a dfs algorithm. Each time it picks a edge with lowest rate.
```

Comment:
The logic `Each time it picks an edge with lowest rate` is not implemented in this function, this function doesn't care about `rate`.